packages/core/mesh/network-manager/src/transport/libdatachannel-transport.ts

```diff
@@ -0,0 +1,241 @@
+
+/**
+ * Transport
+ */
+export class LibDataChannelTransport implements Transport {
+  private _closed = false;
+  readonly closed = new Event();
+  private _connected = false;
+  readonly connected = new Event();
+  readonly errors = new ErrorStream();
+  private readonly _peer: Promise<PeerConnection>;
+  private _channel!: RTCDataChannel;
+  private _stream!: Duplex;
+
+  private readonly _readyForCandidates = new Trigger();
+  private _writeCallback: (() => void) | null = null;
+
+  constructor(private readonly params: LibDataChannelTransportParams) {
+    this._peer = (async () => {
+      const { RTCPeerConnection: PeerConnection } = await import('./datachannel');
+      if (this._closed) {
+        this.errors.raise(Error('connection already closed'));
+      }
+      const peer = new PeerConnection(params.webrtcConfig);
+
+      peer.onicecandidateerror = (event) => {
+        log.error('peer.onicecandidateerror', { event });
+      };
+
+      peer.onconnectionstatechange = (event) => {
+        log.debug('peer.onconnectionstatechange', {
+          event,
+          peerConnectionState: peer.connectionState,
+          transportConnectionState: this._connected,
+        });
+        // TODO(nf): throw error if datachannel does not connect after some time?
+      };
+
+      peer.onicecandidate = async (event) => {
+        log.debug('peer.onicecandidate', { event });
+        if (event.candidate) {
+          await params.sendSignal({
+            payload: {
+              data: {
+                type: 'candidate',
+                candidate: {
+                  candidate: event.candidate.candidate,
+                  // these fields never seem to be not null, but connecting to Chrome doesn't work if they are
+                  sdpMLineIndex: event.candidate.sdpMLineIndex ?? 0,
+                  sdpMid: event.candidate.sdpMid ?? 0,
+                },
+              },
+            },
+          });
+        }
+      };
+      if (params.initiator) {
+        peer
+          .createOffer()
+          .then(async (offer) => {
+            if (peer.connectionState !== 'connecting') {
+              log.error('i am initiator but peer not in state connecting', { peer });
+              this.errors.raise(Error('invalid state: peer is initiator, but other peer not in state connecting'));
+            }
+            log.debug(`im the initiator, creating offer, peer is in state ${peer.connectionState}`, { offer });
+            await peer.setLocalDescription(offer);
+            await params.sendSignal({ payload: { data: { type: offer.type, sdp: offer.sdp } } });
+            return offer;
+          })
+          .catch((err) => {
+            this.errors.raise(err);
+          });
+
+        // TODO(nf): better pattern for adding event handlers to existing object?
+        this._channel = this._channelHandler(peer.createDataChannel('dxos.mesh.transport'));
+        log.info('created data channel', { label: this._channel.label });
```

Comment:
info logs are enabled by default - I think we wanna downgrade this to DEBUG level before merging